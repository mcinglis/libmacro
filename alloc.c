
// Copyright 2014  Malcolm Inglis <http://minglis.id.au>
//
// This file is part of Libmacro.
//
// Libmacro is free software: you can redistribute it and/or modify it under
// the terms of the GNU Affero General Public License as published by the
// Free Software Foundation, either version 3 of the License, or (at your
// option) any later version.
//
// Libmacro is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
// FITNESS FOR A PARTICULAR PURPOSE. See the GNU Affero General Public
// License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with Libmacro. If not, see <https://gnu.org/licenses/>.


#include "alloc.h"

#include <stdlib.h>     // malloc
#include <string.h>     // memcpy


void *
malloc_copy( size_t const size,
             void const * const source )
{
    if ( size == 0 ) {
        return NULL;
    }
    void * const new = malloc( size );
    if ( new == NULL ) {
        return NULL;
    }
    if ( source != NULL ) {
        memcpy( new, source, size );
    }
    return new;
}


