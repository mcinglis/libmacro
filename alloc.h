
// Copyright 2015  Malcolm Inglis <http://minglis.id.au>
//
// This file is part of Libmacro.
//
// Libmacro is free software: you can redistribute it and/or modify it under
// the terms of the GNU Affero General Public License as published by the
// Free Software Foundation, either version 3 of the License, or (at your
// option) any later version.
//
// Libmacro is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
// FITNESS FOR A PARTICULAR PURPOSE. See the GNU Affero General Public
// License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with Libmacro. If not, see <https://gnu.org/licenses/>.


#ifndef LIBMACRO_ALLOC_H
#define LIBMACRO_ALLOC_H


#include <stdlib.h>             // free

#include <libtypes/types.h>     // size_t


// @public
// Takes a type name, and allocates and returns a copy of a compound
// literal of that type, containing the variable arguments as fields.
// This can be used for allocating a primitive value, struct value, or
// an array:
//
//      int * const x = ALLOC( int, 8 );
//      Widget * const w = ALLOC( Widget, .foo = 0xA3, .bar = "wow" );
//      long * const xs = ALLOC( long[ 5 ], 42, 13, 18 );
//
// Note that the unspecified struct fields and array elements will be
// zeroed.
#define ALLOC( TYPE, ... ) \
    malloc_copy( sizeof ( TYPE ), &( TYPE ){ __VA_ARGS__ } )


// Allocates `size` bytes, copies `size` bytes from `source` into that
// memory, and returns a pointer to that memory. Returns `NULL` if
// `size == 0` or if the memory allocation failed. If `source == NULL`,
// nothing is copied, so this acts just like `malloc( size )`.
void * malloc_copy( size_t size, void const * source );


// @public
// Takes a variable number of pointers to allocated memory, and `free()`s
// each one individually.
#define FREE( ... ) \
    PP_MAP( free, PP_SEP_SEMICOLON, __VA_ARGS__ )


#endif

