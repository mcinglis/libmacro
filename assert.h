
// Copyright 2015  Malcolm Inglis <http://minglis.id.au>
//
// This file is part of Libmacro.
//
// Libmacro is free software: you can redistribute it and/or modify it under
// the terms of the GNU Affero General Public License as published by the
// Free Software Foundation, either version 3 of the License, or (at your
// option) any later version.
//
// Libmacro is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
// FITNESS FOR A PARTICULAR PURPOSE. See the GNU Affero General Public
// License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with Libmacro. If not, see <https://gnu.org/licenses/>.


#ifndef LIBMACRO_ASSERT_H
#define LIBMACRO_ASSERT_H


#include <assert.h>

#include <libpp/map.h>          // PP_MAP
#include <libpp/separators.h>   // PP_SEP_SEMICOLON


#ifdef NO_ASSERT
    #define ASSERT( ... )
#else

    // @public
    // Takes a variable number of scalar expressions, and `assert`s each
    // one individually. This provides better assertion error reporting than
    // techniques like `assert( e1 && e2 && ... )`.
    #define ASSERT( ... ) \
        PP_MAP( assert, PP_SEP_SEMICOLON, __VA_ARGS__ )

#endif


#endif


