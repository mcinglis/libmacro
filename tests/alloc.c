

#include <stdio.h>
#include <string.h>

#include "../alloc.h"      // ALLOC
#include "../assert.h"     // ASSERT


typedef struct {
    int foo;
    char const * bar;
    void * baz;
} Widget;


int main( void )
{
    printf( "Allocating int...\n" );
    int * const i = ALLOC( int, 8 );
    ASSERT( i != NULL, *i == 8 );

    printf( "Allocating Widget...\n" );
    Widget * const w = ALLOC( Widget, .foo = 0xA3, .bar = "wow" );
    ASSERT( w != NULL, w->foo == 0xA3, strcmp( w->bar, "wow" ) == 0 );

    printf( "Allocating long array...\n" );
    long * const xs = ALLOC( long[ 6 ], 383, 87, -2983 );
    ASSERT( xs != NULL, xs[ 0 ] == 383, xs[ 1 ] == 87, xs[ 2 ] == -2983,
                        xs[ 3 ] == 0,   xs[ 4 ] == 0,  xs[ 5 ] == 0 );

    FREE( i, w, xs );

    printf( "ALLOC() tests passed.\n" );
}


