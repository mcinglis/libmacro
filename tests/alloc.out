Allocating int...
Allocating Widget...
Allocating long array...
ALLOC() tests passed.
